#### Data set cleaning
**Dataset:** KaggleV2-May-2016.zip  
**Author:** Luis Sergio Pastrana Lemus  
**Date:** 2025-05-24

## Libraries

In [1]:
import os
import sys
import pprint
from pathlib import Path
import pandas as pd

# Define project root dynamically, gets the current directory from whick the notebook belongs and moves one level upper
project_root = Path.cwd().parent

# Add src to sys.path if it is not already
if str(project_root) not in sys.path:
    sys.path.append(str(project_root))

# Import function directly (more controlled than import *)
from src import *

## Path to Data file

In [2]:
# Build route to data file and upload

data_file_path = project_root / "data" / "raw" / "KaggleV2-May-2016.zip"

df_patients = load_dataset_from_zip(data_file_path, "KaggleV2-May-2016.csv", sep='|', header='infer', keep_default_na=False)

###### `LSPL`
###### `"keep_default_na=False" is used to later convert missing values to "pd.NA". This is beneficial because "pd.NA" provides:`

###### `- consistency across data types`

###### `- type integrity preservation`

###### `- cleaner logical operations`

###### `- improved control over missing data.`

###### `Since high performance or heavy computation is not required here, using "pd.NA" is appropriate.`

In [3]:
format_notebook()

## Data set cleaning

In [4]:
df_patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Patient Id       110527 non-null  float64
 1   Appointment ID   110527 non-null  int64  
 2   Gender           110527 non-null  object 
 3   Scheduled Day    110527 non-null  object 
 4   Appointment Day  110527 non-null  object 
 5   Age              110527 non-null  int64  
 6   Neighbourhood    110527 non-null  object 
 7   Scholarship      110527 non-null  int64  
 8   Hipertension     110527 non-null  int64  
 9   Diabetes         110527 non-null  int64  
 10  Alcoholism       110527 non-null  int64  
 11  Handcap          110527 non-null  int64  
 12  SMS_received     110527 non-null  int64  
 13  No-show          110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


##### Missing values

In [5]:
check_existing_missing_values(df_patients, "Data set Patients")

> Dataframe: Data set Patients

*** Warning ***   > No missing values in 'Gender' found
*** Warning ***   > No missing values in 'Scheduled Day' found
*** Warning ***   > No missing values in 'Appointment Day' found
*** Warning ***   > No missing values in 'Neighbourhood' found
*** Warning ***   > No missing values in 'No-show' found



##### Explicit duplicates

In [6]:
print(f"> Number of explicit duplicates in Data frame: {df_patients.duplicated().sum()}")

> Number of explicit duplicates in Data frame: 0


##### Data format

In [7]:
df_patients.columns = normalize_headers_string_format(df_patients.columns)

In [8]:
normalize_df_string_format(df_patients)

patient_id  appointment_id gender         scheduled_day       appointment_day  age      neighbourhood  scholarship  hipertension  \
0       2.987250e+13         5642903      f  2016_04_29t18_38_08z  2016_04_29t00_00_00z   62    jardim_da_penha            0             1   
1       5.589978e+14         5642503      m  2016_04_29t16_08_27z  2016_04_29t00_00_00z   56    jardim_da_penha            0             0   
2       4.262962e+12         5642549      f  2016_04_29t16_19_04z  2016_04_29t00_00_00z   62      mata_da_praia            0             0   
3       8.679512e+11         5642828      f  2016_04_29t17_29_31z  2016_04_29t00_00_00z    8  pontal_de_camburi            0             0   
4       8.841186e+12         5642494      f  2016_04_29t16_07_23z  2016_04_29t00_00_00z   56    jardim_da_penha            0             1   
...              ...             ...    ...                   ...                   ...  ...                ...          ...           ...   
110522  2.572134e+12         5651768      f  2016_05_03t09_15_35z  2016_06_07t00_00_00z   56        maria_ortiz            0             0   
110523  3.596266e+12         5650093      f  2016_05_03t07_27_33z  2016_06_07t00_00_00z   51        maria_ortiz            0             0   
110524  1.557663e+13         5630692      f  2016_04_27t16_03_52z  2016_06_07t00_00_00z   21        maria_ortiz            0             0   
110525  9.213493e+13         5630323      f  2016_04_27t15_09_23z  2016_06_07t00_00_00z   38        maria_ortiz            0             0   
110526  3.775115e+14         5629448      f  2016_04_27t13_30_56z  2016_06_07t00_00_00z   54        maria_ortiz            0             0   

        diabetes  alcoholism  handcap  sms_received no_show  
0              0           0        0             0      no  
1              0           0        0             0      no  
2              0           0        0             0      no  
3              0           0        0             0      no  
4              1           0        0             0      no  
...          ...         ...      ...           ...     ...  
110522         0           0        0             1      no  
110523         0           0        0             1      no  
110524         0           0        0             1      no  
110525         0           0        0             1      no  
110526         0           0        0             1      no  

[110527 rows x 14 columns]

## Implicit duplicates

In [9]:
detect_implicit_duplicates(df_patients)

> Column: 'gender'


 Searching implicit values for: 'gender': 100%|██████████| 2/2 [00:00<00:00, 37.44it/s]


> Column: 'neighbourhood'


 Searching implicit values for: 'neighbourhood': 100%|██████████| 27/27 [00:00<00:00, 77.91it/s]


> Column: 'no_show'


 Searching implicit values for: 'no_show': 100%|██████████| 2/2 [00:00<00:00, 41.17it/s]


## Standardizing boolean values

In [10]:
convert_integer_to_boolean(df_patients, include=['scholarship', 'hipertension', 'diabetes', 'alcoholism', 'handcap', 'sms_received'])

patient_id  appointment_id gender         scheduled_day       appointment_day  age      neighbourhood  scholarship  hipertension  \
0       2.987250e+13         5642903      f  2016_04_29t18_38_08z  2016_04_29t00_00_00z   62    jardim_da_penha        False          True   
1       5.589978e+14         5642503      m  2016_04_29t16_08_27z  2016_04_29t00_00_00z   56    jardim_da_penha        False         False   
2       4.262962e+12         5642549      f  2016_04_29t16_19_04z  2016_04_29t00_00_00z   62      mata_da_praia        False         False   
3       8.679512e+11         5642828      f  2016_04_29t17_29_31z  2016_04_29t00_00_00z    8  pontal_de_camburi        False         False   
4       8.841186e+12         5642494      f  2016_04_29t16_07_23z  2016_04_29t00_00_00z   56    jardim_da_penha        False          True   
...              ...             ...    ...                   ...                   ...  ...                ...          ...           ...   
110522  2.572134e+12         5651768      f  2016_05_03t09_15_35z  2016_06_07t00_00_00z   56        maria_ortiz        False         False   
110523  3.596266e+12         5650093      f  2016_05_03t07_27_33z  2016_06_07t00_00_00z   51        maria_ortiz        False         False   
110524  1.557663e+13         5630692      f  2016_04_27t16_03_52z  2016_06_07t00_00_00z   21        maria_ortiz        False         False   
110525  9.213493e+13         5630323      f  2016_04_27t15_09_23z  2016_06_07t00_00_00z   38        maria_ortiz        False         False   
110526  3.775115e+14         5629448      f  2016_04_27t13_30_56z  2016_06_07t00_00_00z   54        maria_ortiz        False         False   

        diabetes  alcoholism  handcap  sms_received no_show  
0          False       False    False         False      no  
1          False       False    False         False      no  
2          False       False    False         False      no  
3          False       False    False         False      no  
4           True       False    False         False      no  
...          ...         ...      ...           ...     ...  
110522     False       False    False          True      no  
110523     False       False    False          True      no  
110524     False       False    False          True      no  
110525     False       False    False          True      no  
110526     False       False    False          True      no  

[110527 rows x 14 columns]

## Standardize gender values

In [11]:
standardize_gender_values(df_patients, include=['gender'])

patient_id  appointment_id  gender         scheduled_day       appointment_day  age      neighbourhood  scholarship  hipertension  \
0       2.987250e+13         5642903  female  2016_04_29t18_38_08z  2016_04_29t00_00_00z   62    jardim_da_penha        False          True   
1       5.589978e+14         5642503    male  2016_04_29t16_08_27z  2016_04_29t00_00_00z   56    jardim_da_penha        False         False   
2       4.262962e+12         5642549  female  2016_04_29t16_19_04z  2016_04_29t00_00_00z   62      mata_da_praia        False         False   
3       8.679512e+11         5642828  female  2016_04_29t17_29_31z  2016_04_29t00_00_00z    8  pontal_de_camburi        False         False   
4       8.841186e+12         5642494  female  2016_04_29t16_07_23z  2016_04_29t00_00_00z   56    jardim_da_penha        False          True   
...              ...             ...     ...                   ...                   ...  ...                ...          ...           ...   
110522  2.572134e+12         5651768  female  2016_05_03t09_15_35z  2016_06_07t00_00_00z   56        maria_ortiz        False         False   
110523  3.596266e+12         5650093  female  2016_05_03t07_27_33z  2016_06_07t00_00_00z   51        maria_ortiz        False         False   
110524  1.557663e+13         5630692  female  2016_04_27t16_03_52z  2016_06_07t00_00_00z   21        maria_ortiz        False         False   
110525  9.213493e+13         5630323  female  2016_04_27t15_09_23z  2016_06_07t00_00_00z   38        maria_ortiz        False         False   
110526  3.775115e+14         5629448  female  2016_04_27t13_30_56z  2016_06_07t00_00_00z   54        maria_ortiz        False         False   

        diabetes  alcoholism  handcap  sms_received no_show  
0          False       False    False         False      no  
1          False       False    False         False      no  
2          False       False    False         False      no  
3          False       False    False         False      no  
4           True       False    False         False      no  
...          ...         ...      ...           ...     ...  
110522     False       False    False          True      no  
110523     False       False    False          True      no  
110524     False       False    False          True      no  
110525     False       False    False          True      no  
110526     False       False    False          True      no  

[110527 rows x 14 columns]

## Convert to integer patient_id column

In [12]:
convert_ndtype_to_numeric(df_patients, include=['patient_id'], type='integer')


*** Warning ***   > Numeric values that are not whole integer found in column [patient_id]:
3950       93779.52927
73228     537615.28476
73303     141724.16655
100517     39217.84439
105430     43741.75652
Name: patient_id, dtype: float64

> Conversion unsuccessful, numeric values non whole integer amount: 5



patient_id  appointment_id  gender         scheduled_day       appointment_day  age      neighbourhood  scholarship  hipertension  \
0       2.987250e+13         5642903  female  2016_04_29t18_38_08z  2016_04_29t00_00_00z   62    jardim_da_penha        False          True   
1       5.589978e+14         5642503    male  2016_04_29t16_08_27z  2016_04_29t00_00_00z   56    jardim_da_penha        False         False   
2       4.262962e+12         5642549  female  2016_04_29t16_19_04z  2016_04_29t00_00_00z   62      mata_da_praia        False         False   
3       8.679512e+11         5642828  female  2016_04_29t17_29_31z  2016_04_29t00_00_00z    8  pontal_de_camburi        False         False   
4       8.841186e+12         5642494  female  2016_04_29t16_07_23z  2016_04_29t00_00_00z   56    jardim_da_penha        False          True   
...              ...             ...     ...                   ...                   ...  ...                ...          ...           ...   
110522  2.572134e+12         5651768  female  2016_05_03t09_15_35z  2016_06_07t00_00_00z   56        maria_ortiz        False         False   
110523  3.596266e+12         5650093  female  2016_05_03t07_27_33z  2016_06_07t00_00_00z   51        maria_ortiz        False         False   
110524  1.557663e+13         5630692  female  2016_04_27t16_03_52z  2016_06_07t00_00_00z   21        maria_ortiz        False         False   
110525  9.213493e+13         5630323  female  2016_04_27t15_09_23z  2016_06_07t00_00_00z   38        maria_ortiz        False         False   
110526  3.775115e+14         5629448  female  2016_04_27t13_30_56z  2016_06_07t00_00_00z   54        maria_ortiz        False         False   

        diabetes  alcoholism  handcap  sms_received no_show  
0          False       False    False         False      no  
1          False       False    False         False      no  
2          False       False    False         False      no  
3          False       False    False         False      no  
4           True       False    False         False      no  
...          ...         ...      ...           ...     ...  
110522     False       False    False          True      no  
110523     False       False    False          True      no  
110524     False       False    False          True      no  
110525     False       False    False          True      no  
110526     False       False    False          True      no  

[110527 rows x 14 columns]

In [13]:
df_patients['patient_id'] = df_patients['patient_id'].astype('int')
print(df_patients.sample(25, random_state=333))
      

             patient_id  appointment_id  gender         scheduled_day       appointment_day  age      neighbourhood  scholarship  hipertension  \
46995    62566927247961         5607061  female  2016_04_20t11_02_15z  2016_05_18t00_00_00z    1             jabour        False         False   
82570     3996739777887         5645533    male  2016_05_02t10_03_24z  2016_05_02t00_00_00z   11          conquista        False         False   
71800    74981215322154         5664902  female  2016_05_05t13_39_17z  2016_05_05t00_00_00z   66        santa_lúcia        False         False   
78525   746757565862383         5700616  female  2016_05_16t09_37_32z  2016_05_17t00_00_00z   78        santo_andre        False          True   
63501    49375648372849         5704393  female  2016_05_16t15_37_53z  2016_05_19t00_00_00z   46           da_penha        False         False   
38007    85922122763292         5634129    male  2016_04_28t10_25_11z  2016_05_25t00_00_00z   39     forte_são_joão         

## Standardizing date/time values

In [14]:
replace_string_values_datetime(df_patients, include=['scheduled_day', 'appointment_day'], frmt="%Y_%m_%dT%H_%M_%SZ")

patient_id  appointment_id  gender             scheduled_day           appointment_day  age      neighbourhood  scholarship  \
0        29872499824296         5642903  female 2016-04-29 18:38:08+00:00 2016-04-29 00:00:00+00:00   62    jardim_da_penha        False   
1       558997776694438         5642503    male 2016-04-29 16:08:27+00:00 2016-04-29 00:00:00+00:00   56    jardim_da_penha        False   
2         4262962299951         5642549  female 2016-04-29 16:19:04+00:00 2016-04-29 00:00:00+00:00   62      mata_da_praia        False   
3          867951213174         5642828  female 2016-04-29 17:29:31+00:00 2016-04-29 00:00:00+00:00    8  pontal_de_camburi        False   
4         8841186448183         5642494  female 2016-04-29 16:07:23+00:00 2016-04-29 00:00:00+00:00   56    jardim_da_penha        False   
...                 ...             ...     ...                       ...                       ...  ...                ...          ...   
110522    2572134369293         5651768  female 2016-05-03 09:15:35+00:00 2016-06-07 00:00:00+00:00   56        maria_ortiz        False   
110523    3596266328735         5650093  female 2016-05-03 07:27:33+00:00 2016-06-07 00:00:00+00:00   51        maria_ortiz        False   
110524   15576631729893         5630692  female 2016-04-27 16:03:52+00:00 2016-06-07 00:00:00+00:00   21        maria_ortiz        False   
110525   92134931435557         5630323  female 2016-04-27 15:09:23+00:00 2016-06-07 00:00:00+00:00   38        maria_ortiz        False   
110526  377511518121127         5629448  female 2016-04-27 13:30:56+00:00 2016-06-07 00:00:00+00:00   54        maria_ortiz        False   

        hipertension  diabetes  alcoholism  handcap  sms_received no_show  
0               True     False       False    False         False      no  
1              False     False       False    False         False      no  
2              False     False       False    False         False      no  
3              False     False       False    False         False      no  
4               True      True       False    False         False      no  
...              ...       ...         ...      ...           ...     ...  
110522         False     False       False    False          True      no  
110523         False     False       False    False          True      no  
110524         False     False       False    False          True      no  
110525         False     False       False    False          True      no  
110526         False     False       False    False          True      no  

[110527 rows x 14 columns]

## Statistical information on quantitative and categorical values

In [15]:
print(df_patients.describe())

         patient_id  appointment_id            age
count  1.105270e+05    1.105270e+05  110527.000000
mean   1.474963e+14    5.675305e+06      37.088874
std    2.560949e+14    7.129575e+04      23.110205
min    3.921700e+04    5.030230e+06      -1.000000
25%    4.172614e+12    5.640286e+06      18.000000
50%    3.173184e+13    5.680573e+06      37.000000
75%    9.439172e+13    5.725524e+06      55.000000
max    9.999816e+14    5.790484e+06     115.000000


In [16]:
print(df_patients.describe(include='object'))

        gender   neighbourhood no_show
count   110527          110527  110527
unique       2              81       2
top     female  jardim_camburi      no
freq     71840            7717   88208


In [17]:
print(df_patients['age'].value_counts())

age
 0      3539
 1      2273
 52     1746
 49     1652
 53     1651
        ... 
 115       5
 100       4
 102       2
 99        1
-1         1
Name: count, Length: 104, dtype: int64


In [18]:
Q1 = df_patients['age'].quantile(0.25)
Q3 = df_patients['age'].quantile(0.75)
IQR = Q3 - Q1

outliers = df_patients[(df_patients['age'] < Q1 - 1.5 * IQR) | (df_patients['age'] > Q3 + 1.5 * IQR)]
print(outliers)

            patient_id  appointment_id  gender             scheduled_day           appointment_day  age neighbourhood  scholarship  hipertension  \
63912   31963211613981         5700278  female 2016-05-16 09:17:44+00:00 2016-05-19 00:00:00+00:00  115    andorinhas        False         False   
63915   31963211613981         5700279  female 2016-05-16 09:17:44+00:00 2016-05-19 00:00:00+00:00  115    andorinhas        False         False   
68127   31963211613981         5562812  female 2016-04-08 14:29:17+00:00 2016-05-16 00:00:00+00:00  115    andorinhas        False         False   
76284   31963211613981         5744037  female 2016-05-30 09:44:51+00:00 2016-05-30 00:00:00+00:00  115    andorinhas        False         False   
97666  748234579244724         5717451  female 2016-05-19 07:57:56+00:00 2016-06-03 00:00:00+00:00  115      são_jose        False          True   

       diabetes  alcoholism  handcap  sms_received no_show  
63912     False       False     True         False

In [19]:
media = df_patients['age'].mean()
std = df_patients['age'].std()

outliers = df_patients[(df_patients['age'] < media - 3 * std) | (df_patients['age'] > media + 3 * std)]
print(outliers)

            patient_id  appointment_id  gender             scheduled_day           appointment_day  age neighbourhood  scholarship  hipertension  \
63912   31963211613981         5700278  female 2016-05-16 09:17:44+00:00 2016-05-19 00:00:00+00:00  115    andorinhas        False         False   
63915   31963211613981         5700279  female 2016-05-16 09:17:44+00:00 2016-05-19 00:00:00+00:00  115    andorinhas        False         False   
68127   31963211613981         5562812  female 2016-04-08 14:29:17+00:00 2016-05-16 00:00:00+00:00  115    andorinhas        False         False   
76284   31963211613981         5744037  female 2016-05-30 09:44:51+00:00 2016-05-30 00:00:00+00:00  115    andorinhas        False         False   
97666  748234579244724         5717451  female 2016-05-19 07:57:56+00:00 2016-06-03 00:00:00+00:00  115      são_jose        False          True   

       diabetes  alcoholism  handcap  sms_received no_show  
63912     False       False     True         False

###### `LSPL`

###### `For those patients identified as outliers (5) with 115 years of age, 2 of them did appear, the other three being only (3) and not giving indications of being registration errors in the system, will be kept in the dataset.`

##### Female patient with age (-1)

In [20]:
df_patients_negative_age = df_patients.loc[(df_patients['age'] == -1), :]
print(df_patients_negative_age)

            patient_id  appointment_id  gender             scheduled_day           appointment_day  age neighbourhood  scholarship  hipertension  \
99832  465943158731293         5775010  female 2016-06-06 08:58:13+00:00 2016-06-06 00:00:00+00:00   -1         romão        False         False   

       diabetes  alcoholism  handcap  sms_received no_show  
99832     False       False    False         False      no  


###### `LSPL`

###### `Since the patient showed up for her appointment, her age was probably an error when entering her information into the system, so her age will be estimated using the mean of the dataset.`



In [21]:
average_age = df_patients['age'].mean()
print(f"> Average patients' age: {average_age:.2f}")

> Average patients' age: 37.09


In [22]:
df_patients.loc[99832, 'age'] =  average_age.astype('int')
print(df_patients.loc[99832, :])

patient_id                   465943158731293
appointment_id                       5775010
gender                                female
scheduled_day      2016-06-06 08:58:13+00:00
appointment_day    2016-06-06 00:00:00+00:00
age                                       37
neighbourhood                          romão
scholarship                            False
hipertension                           False
diabetes                               False
alcoholism                             False
handcap                                False
sms_received                           False
no_show                                   no
Name: 99832, dtype: object


##### Patients with zero age

In [23]:
df_patients_zero_age = df_patients.loc[(df_patients['age'] == 0), :]
print(df_patients_zero_age.sample(25, random_state=333))

             patient_id  appointment_id  gender             scheduled_day           appointment_day  age        neighbourhood  scholarship  \
54493     9841193721753         5703077  female 2016-05-16 13:28:50+00:00 2016-05-16 00:00:00+00:00    0     ilha_do_príncipe        False   
37994      563638288586         5712900  female 2016-05-18 09:27:08+00:00 2016-05-18 00:00:00+00:00    0                romão        False   
59604   438275565697561         5718209  female 2016-05-19 09:05:19+00:00 2016-05-19 00:00:00+00:00    0          santo_andre        False   
70096     6677858368354         5746846    male 2016-05-30 14:37:16+00:00 2016-05-30 00:00:00+00:00    0      jardim_da_penha        False   
37273   894537485372141         5512453    male 2016-03-28 08:53:54+00:00 2016-05-03 00:00:00+00:00    0       forte_são_joão        False   
103979  566647938996295         5786898  female 2016-06-08 09:03:00+00:00 2016-06-08 00:00:00+00:00    0          santo_andre        False   
65676 

In [24]:
print(df_patients_zero_age.describe(include='bool'))

       scholarship hipertension diabetes alcoholism handcap sms_received
count         3539         3539     3539       3539    3539         3539
unique           2            1        1          1       2            2
top          False        False    False      False   False        False
freq          3487         3539     3539       3539    3538         2642


In [25]:
print(df_patients_zero_age.describe(include='object'))

       gender neighbourhood no_show
count    3539          3539    3539
unique      2            75       2
top      male   maria_ortiz      no
freq     1818           242    2900


###### `LSPL`

###### `Since most patients with age (0) have information indicating that they do not appear to belong to erroneous records, their age will be modified with the average of the dataset.`

In [26]:
df_patients.loc[(df_patients['age'] == 0), 'age'] = average_age.astype('int')

In [27]:
print(df_patients['age'].describe())

count    110527.000000
mean         38.273933
std          22.104720
min           1.000000
25%          20.000000
50%          37.000000
75%          55.000000
max         115.000000
Name: age, dtype: float64


## Generate a clean data file

In [28]:
project_root = Path.cwd().parent
processed_path = project_root / "data" / "processed" / "patients_clean.csv"

df_patients.to_csv(processed_path, index=False)